tsl/src/continuous_aggs/planner.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include <postgres.h>
+
+#include <nodes/makefuncs.h>
+#include <nodes/nodeFuncs.h>
+#include <nodes/nodes.h>
+#include <nodes/pg_list.h>
+#include <parser/parse_func.h>
+
+#include "continuous_aggs/common.h"
+#include "planner.h"
+#include "ts_catalog/continuous_aggs_watermark.h"
+
+/*
+ * The watermark function of a CAgg query is embedded into further functions. It
+ * has the following structure:
+ *
+ * 1. a coalesce expression
+ * 2. an optional to timestamp conversion function (date, timestamp, timestamptz)
+ * 3. the actual watermark function
+ *
+ * For example:
+ *        ... COALESCE(to_timestamp(cagg_watermark(59)), XXX) ...
+ *        ... COALESCE(cagg_watermark(59), XXX) ...
+ *
+ * We use the following data structure while walking to the query to analyze the query
+ * and collect references to the needed functions. The data structure contains:
+ *
+ *   (a) values (e.g., function Oids) which are needed to analyze the query
+ *   (b) values that are changed during walking through the query
+ *       (e.g., references to parent functions)
+ *   (c) result data like the watermark functions and their parent functions
+ */
+typedef struct
+{
+	/* (a) Values initialized after creating the context */
+	Oid watermark_function_oid;	  // Oid of the watermark function
+	List *to_timestamp_func_oids; // List of Oids of the timestamp conversion functions
+
+	/* (b) Values changed while walking through the query */
+	CoalesceExpr *parent_coalesce_expr; // the current parent coalesce_expr
+	FuncExpr *parent_to_timestamp_func; // the current parent timestamp function
+
+	/* (c) Result values */
+	List *watermark_parent_functions; // List of parent functions of a watermark (1) and (2)
+	List *watermark_functions;		  // List of watermark functions (3)
+	List *relids;					  // List of used relids by the query
+	bool valid_query;				  // Is the query valid a valid CAgg query or not
+} ConstifyWatermarkContext;
+
+/*
+ * Walk through the elements of the query and detect the watermark functions and their
+ * parent functions.
+ */
+static bool
+constify_cagg_watermark_walker(Node *node, ConstifyWatermarkContext *context)
```

Comment:
I don't know much about cagg planning, so a pretty basic question -- where do we add these conditions for the cagg watermark in the first place? Maybe it's possible to have this logic there? Now it looks as if we're first adding the conditions and then parsing them back.

Reply:
When a CAgg is created internally we create a VIEW and if the cagg is realtime this view is a UNION view querying the original hypertable (or cagg in case of hierarchical) and the materialization hypertable. To this UNION view we add the predicates using cagg_watermark to read data before and after the watermark.